nixos/modules/tasks/filesystems/overlayfs.nix

```diff
@@ -0,0 +1,147 @@
+        default = null;
+        description = lib.mdDoc ''
+          The path to the upperdir.
+
+          If this is null, a read-only overlay is created using the lowerdir.
+
+          If you set this to some value you MUST also set `workdir`.
+        '';
+      };
+
+      workdir = lib.mkOption {
+        type = lib.types.nullOr lib.types.path;
+        default = null;
+        description = lib.mdDoc ''
+          The path to the workdir.
+
+          This MUST be set if you set `upperdir`.
+        '';
+      };
+
+    };
+
+    config = lib.mkIf (config.overlay.lowerdir != null) {
+      fsType = "overlay";
+      device = lib.mkDefault "overlay";
+
+      options =
+        let
+          rwService = "rw-${utils.escapeSystemdPath config.mountPoint}.service";
+
+          prefix = sysrootPrefix config;
+
+          lowerdir = map (s: prefix + s) config.overlay.lowerdir;
+          upperdir = prefix + config.overlay.upperdir;
+          workdir = prefix + config.overlay.workdir;
+        in
+        [
+          "x-systemd.wanted-by=local-fs.target"
+          "x-systemd.before=local-fs.target"
+          "lowerdir=${lib.concatStringsSep ":" lowerdir}"
+        ] ++ lib.optionals (config.overlay.upperdir != null) [
+          "upperdir=${upperdir}"
+          "workdir=${workdir}"
+          "x-systemd.requires=${rwService}"
+          "x-systemd.after=${rwService}"
+        ] ++ (map (s: "x-systemd.requires-mounts-for=${s}") lowerdir);
+    };
+
+  };
+in
+
+{
+
+  options = {
+
+    # Merge the overlay options into the fileSystems option.
+    fileSystems = lib.mkOption {
+      type = lib.types.attrsOf (lib.types.submodule [ overlayOpts ]);
+    };
+
+  };
+
+  config =
+    let
+      overlayFileSystems = lib.filterAttrs (_name: fs: (fs.overlay.lowerdir != null)) config.fileSystems;
+      initrdFileSystems = lib.filterAttrs (_name: utils.fsNeededForBoot) overlayFileSystems;
+      userspaceFileSystems = lib.filterAttrs (_name: fs: (!utils.fsNeededForBoot fs)) overlayFileSystems;
+    in
+    {
+
+      boot.initrd.availableKernelModules = lib.mkIf (initrdFileSystems != { }) [ "overlay" ];
+
+      assertions = lib.concatLists (lib.mapAttrsToList
+        (_: fs: [
+          {
+            assertion = (fs.overlay.upperdir == null) == (fs.overlay.workdir == null);
+            message = "You cannot define a `lowerdir` without a `workdir` and vice versa for mount point: ${fs.mountPoint}";
+          }
+          {
+            assertion = (fs.overlay.lowerdir != null && fs.overlay.upperdir == null) -> (lib.length fs.overlay.lowerdir) >= 2;
+            message = "A read only overlay (without an `upperdir`) requires at least 2 `lowerdir`s.";
+          }
+        ])
+        config.fileSystems);
+
+      boot.initrd.systemd.services = lib.mkMerge (lib.mapAttrsToList preMountService initrdFileSystems);
+      systemd.services = lib.mkMerge (lib.mapAttrsToList preMountService userspaceFileSystems);
```

Comment:
﻿nikstur: ok I figured out the problem... sorta.

`systemd[1]: local-fs.target: Cannot add dependency job to transaction, deleting job local-fs.target/start again: Unit rw-initrd\x2doverlay.service not found.`

So the problem is that the initrd-overlay's `x-systemd.requires=rw-initrd\x2doverlay.service` doesn't exist in stage 2. Which of course it doesn't; we only make that dep in stage 1 because that's what it's for. But the `Requires=` dependency means that, even though `/initrd-overlay` is already mounted, it still causes `local-fs.target` to be un-start-able. A `Requires=foo.service` dependency means that you don't get started at all if `foo.service` doesn't exist.

The (possibly buggy?) systemd behavior is that _some_ of `local-fs.target`'s dependencies are getting pulled in anyway, but not all of them. So `/userspace-overlay` is being pulled in while `/ro-userspace-overlay` is not. I _think_ the `x-systemd.wanted-by=local-fs.target` option is "fixing" it because it's unintentionally taking advantage of this inconsistency and resulting in `/ro-userspace-overlay` becoming one of the lucky ones that gets pulled in anyway. I'm not exactly sure why systemd is pulling in *some* of these dependencies but not all of them, but regardless, the solution is that we need to not be failing `local-fs.target` with broken `Requires=` dependencies.

If we add the rw service to stage 2 as well, the issue goes away. This could arguably be seen as a good idea anyway, as it more accurately reflects the dependencies in stage 2; I dunno. The alternatives are 1) we could make the service required only by the initrd mount and not the stage 2 mount by adding `requiredBy` to the service instead of `x-systemd.requires` to the mount, or 2) we could weaken the dependency to `x-systemd.wants=${rwService}`

```diff
diff --git a/nixos/modules/tasks/filesystems/overlayfs.nix b/nixos/modules/tasks/filesystems/overlayfs.nix
index 0a08bcc2c136..2a37e398d51e 100644
--- a/nixos/modules/tasks/filesystems/overlayfs.nix
+++ b/nixos/modules/tasks/filesystems/overlayfs.nix
@@ -91,8 +91,6 @@ let
           workdir = prefix + config.overlay.workdir;
         in
         [
-          "x-systemd.wanted-by=local-fs.target"
-          "x-systemd.before=local-fs.target"
           "lowerdir=${lib.concatStringsSep ":" lowerdir}"
         ] ++ lib.optionals (config.overlay.upperdir != null) [
           "upperdir=${upperdir}"
@@ -120,7 +118,6 @@ in
     let
       overlayFileSystems = lib.filterAttrs (_name: fs: (fs.overlay.lowerdir != null)) config.fileSystems;
       initrdFileSystems = lib.filterAttrs (_name: utils.fsNeededForBoot) overlayFileSystems;
-      userspaceFileSystems = lib.filterAttrs (_name: fs: (!utils.fsNeededForBoot fs)) overlayFileSystems;
     in
     {
 
@@ -140,7 +137,7 @@ in
         config.fileSystems);
 
       boot.initrd.systemd.services = lib.mkMerge (lib.mapAttrsToList preMountService initrdFileSystems);
-      systemd.services = lib.mkMerge (lib.mapAttrsToList preMountService userspaceFileSystems);
+      systemd.services = lib.mkMerge (lib.mapAttrsToList preMountService overlayFileSystems);
 
     };
 
```

Reply:
Thank you for digging! Both alternatives sound fine. I'll play around with them.

Reply:
Well, there were three options :P There's the patch above, which *I think* is my preferred solution. Or we can make the dependency only exist in stage 1 by putting it in `requiredBy` on the service. Or we can relax the dependency to a `wants`.

Reply:
It turns out there actually is no `x-systemd.wants` field.

I now landed on making the dependency only exist in the initrd. I think this makes the code a little more elegant and the fstab more readable.

The test now seems to work reproducibly!